grafana-agent: ships without web ui

### Describe the bug
I started grafana-agent in flow mode, and tried to access the http API at http://localhost:12345.

Unfortunately, instead of providing a web interface to debug the current list of components, I get a "404 page not found".

### Steps To Reproduce
Steps to reproduce the behavior:
1. Run grafana-agent in flow mode.
2. Access port 12345 via HTTP

### Expected behavior
I'd expect grafana-agent to ship with the web interface.

### Screenshots
-

### Additional context
-

### Notify maintainers
@emilylange @flokli 


### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

It looks like `web/ui` is what's building the frontend. Unfortunately, the web interface seems managed by yarn (and has a yarn.lock), so we'd need to first provide a `package-lock.json`.

I tried to have `npm` resolve the dependency, but it seems it's not able to:

```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: grafana-agent@undefined
npm ERR! Found: react@18.2.0
npm ERR! node_modules/react
npm ERR!   react@"^18.2.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer react@"^16.8.0 || ^17.0.0" from @grafana/ui@9.5.2
npm ERR! node_modules/@grafana/ui
npm ERR!   @grafana/ui@"^9.5.2" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! 
npm ERR! For a full report see:
npm ERR! /home/flokli/.npm/_logs/2023-08-02T09_11_41_196Z-eresolve-report.txt

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/flokli/.npm/_logs/2023-08-02T09_11_41_196Z-debug-0.log
```

--- a/pkgs/servers/monitoring/grafana-agent/default.nix
+++ b/pkgs/servers/monitoring/grafana-agent/default.nix
@@ -1,11 +1,15 @@
 { lib
 , buildGoModule
 , fetchFromGitHub
+, fetchYarnDeps
+, fixup_yarn_lock
 , grafana-agent
 , nixosTests
+, nodejs
 , stdenv
 , systemd
 , testers
+, yarn
 }:
 
 buildGoModule rec {
@@ -22,6 +26,11 @@ buildGoModule rec {
   vendorHash = "sha256-MqUkGKOzx8Qo9xbD9GdUryVwKjpVUOXFo2x0/2uz8Uk=";
   proxyVendor = true; # darwin/linux hash mismatch
 
+  frontendYarnOfflineCache = fetchYarnDeps {
+    yarnLock = src + "/web/ui/yarn.lock";
+    hash = "sha256-xJEPubIDjlZQL70UGha1MHbeek6KFPaxZGb5IRgMlTA=";
+  };
+
   ldflags = let
     prefix = "github.com/grafana/agent/pkg/build";
   in [
@@ -34,7 +43,10 @@ buildGoModule rec {
     "-X ${prefix}.BuildDate=1980-01-01T00:00:00Z"
   ];
 
+  nativeBuildInputs = [ fixup_yarn_lock nodejs yarn ];
+
   tags = [
+    "builtinassets"
     "nonetwork"
     "nodocker"
     "promtail_journal_enabled"
@@ -43,7 +55,20 @@ buildGoModule rec {
   subPackages = [
     "cmd/grafana-agent"
     "cmd/grafana-agentctl"
+    "web/ui"
   ];
+
+  preBuild = ''
+    export HOME="$TMPDIR"
+
+    pushd web/ui
+    fixup_yarn_lock yarn.lock
+    yarn config --offline set yarn-offline-mirror $frontendYarnOfflineCache
+    yarn install --offline --frozen-lockfile --ignore-platform --ignore-scripts --no-progress --non-interactive
+    patchShebangs node_modules
+    yarn --offline run build
+    popd
+  '';
 
   # uses go-systemd, which uses libsystemd headers
   # https://github.com/coreos/go-systemd/issues/351

